api.go

```diff
@@ -84,30 +88,49 @@ func (d *Device) Query(request *http.Request) (*http.Response, error) {
 		if resp.StatusCode() != http.StatusOK {
 			return nil, fmt.Errorf("failed to create object: %s", resp.Status())
 		}
-		reqEnv.ObjectUrl = createEnv.ObjectUrl
+		reqEnv.ObjectUrl = createEnv.ObjectUrl[0]
 
 		// Upload previously read payload and remaining body
 		reader := io.MultiReader(bytes.NewReader(payload), request.Body)
 
 		// Read one more byte to check if max is crossed
-		b, err := io.ReadAll(io.LimitReader(reader, int64(RequestObjectMax)+1))
-		if err != nil {
-			return nil, err
-		}
-		if len(b) > RequestObjectMax {
-			return nil, fmt.Errorf("payload too large for this device")
-		}
-		reader = bytes.NewReader(b)
+		payloadTotalLength := 0
+		var wg sync.WaitGroup
+		var uploadErr error
+		for i := 0; ; i++ {
+			b, err := io.ReadAll(io.LimitReader(reader, int64(PartSize)))
+			if err != nil {
+				return nil, err
+			}
+			if len(b) == 0 {
+				break
+			}
+			payloadTotalLength += len(b)
+			if payloadTotalLength > RequestObjectMax {
+				return nil, fmt.Errorf("payload too large for this device")
+			}
+
+			wg.Add(1)
+			go func(urlNumber int) {
+				defer wg.Done()
+				hresp, err := resty.New().R().
+					SetBody(bytes.NewReader(b)).
+					SetDoNotParseResponse(true).
+					Put(createEnv.ObjectUrl[urlNumber])
+				if err != nil && uploadErr != nil {
+					uploadErr = err
+				}
+				if hresp.StatusCode() != http.StatusOK {
```

Comment:
> Just curious, I thought we need to keep track of the response. Because the response contains the S3 etags? And the "multipart end" command needs to send all these tags back to S3 to construct the object?

We dont need to keep track of the response. The multipart begin and end transactions take care of that. All we need to do is upload to the urls and then call "multipart end" transaction.